### PR TITLE
MultiBands, and tools improvements (mostly)

### PR DESCRIPTION
**Pull request by ocourtin**
A) MultiBands aka #56
    - Switch from PIL to OpenCV for slippy map images (to allow multibands images handling)

    - SlippyMapTileConcatenation produce now an aggregate C,W,H NumPy tensor rather than a list of RGB images

    - First Unet-like encoder layer, could be adjust upon num_channels (so not anymore harcoded 3 but now 1-N)

    - ONNX export tool was updated consequently

    - Defaults weights (before reusing any ImageNet RGB pre trained weights bands) use Xavier initialization

    - Nothing specific was done at this moment on ResNet mean/stdev values (still to explore if/when needed)

    - Dataset configuration file allow channels configuration for bands on each choosen slippymap dir
      and choosen channels are reported in the train ouputs logs

    - Add an experimental option in export tool, to allow to reduce channels, from a pth to an another pth.
      At this stage, allow to make some experiments, on how to reuse ImageNet weights in a MultiBands/Fusion context.
      This point is still an open subject to find best training approach (i.e rather than train from scratch)

    - Nota: still to explore alternate NN topology to deal with multibands
      for example, where extra bands are copied on each Encoder layers (and not only on the first one, as in this PR):
      cf: https://hal.archives-ouvertes.fr/hal-01523573/document

   Data Augmentation
    - Refactored rotate and flip
    - Removed what looks like an useless crop treatment
    - data_augmentation occurence probability (flip or rotate for now) became an hyperparameter in userland
    - Select best resampling resize interpolation as previously, but now also on downsampling
    - To keep in mind that even some basic DataAugmentation as flip/rotate can't be been performed with PyTorch on GPU
      (https://discuss.pytorch.org/t/torch-from-numpy-not-support-negative-strides/3663)

B) Tools:

 Download:
  - Add a WFS and TMS server type handling (XYZ still the default)
  - Timeout could be specified in userland (some WMS performing pre/post-treatments could take times)
  - Use and improve Log handling (e.g giving more information on already downloaded tiles)

 Cover:
  - Add ability to render a cover from a lat/lon bbox, or from a slippy map directory
    (as a consequence the features parameter become optional)

 Predict:
  - Add an option to generate directly masks (rather than probs)
  - Refactor the tile_buffer provider (slightly improve perfs)
  - model parameter is not mandatory anymore (as we use cuda if CUDA_VISIBLE_DEVICES not empty)

 Compare:
  - Full refactor, Compare tool now allow 3 differents modes:
     - side: Close to the previous behaviour (and so the default), but allow 2 to N images to compare (cf #80)
     - diff: Work with images/masks/labels and compute a single diff image for efficient QoD check
     - list: Render a cover with IoU metric on each mask/label couple tiles

 Tile:
  - Add a first tiling tool implementation, to tile either images or labels from a raster,
    performances are decents for a mono process stuff,
    and it's able to deal with no_data borders (by removing the related tiles)

 Rasterize:
  - Refactor the GeoJSON parser, to cleanly handle all GeoJSON surfacic geometries
    (GeometryCollection, MultiPolygon, Polygon) and N-Dimensional GeoJSON coordinates
    Nota: we still blindly assume GeoJSON coordinates inputs are EPSG:4326
  - Use and improve Log handling

 Predict, Subset, Download, Rasterize, Compare, Masks, Tile:
  - Add a leaflet client generation option, to allow an easy slippy map visual inspection

C) Maintenance:

 Bugfixes
  - Fix a ZeroDiv issue, who was still remaning in mIoU metrics

 Performances
  - Switch from Pillow to Pillow-simd (cf https://github.com/uploadcare/pillow-simd)
    Imply recent Intel/AMD proc, but significant perf improvements on PIL treatment stuff (about x6)
    Nota: didn't see a significative difference for RoboSat use case beetwen SSE4 and AVX2 (i.e SSE4 is 'enough')

 Versions stuff
  - Upgrade rasterio version to the latest stable one (1.0.9)
  - nn.functional.sample -> nn.functional.interpolate (to avoid 0.4.1 related warning)

 Credits
  - echoed my name >> AUTHORS.md

D) Userland considerations:
   - channels configuration now mandatory in dataset.toml (as an array of table)
   - data_augmentation ratio and resnet pretrained as new mandatory hyper-parameters in model.toml
   - cuda bool is no longer needed in model.toml (by default use now all CUDA_VISIBLE_DEVICES available)
   - libwebp and libjpeg are mandatory for pillow-simd (so imply updating install doc)
   - In compare tool, maximum parameter was removed (could't imagine a use case) and minimum was rename to minimum_fg
   - In cover tool the features parameter is not anymore a positional (became an optional)

NOTES:
  - Could you Daniel give me a hint how the robosat unit tests are supposed to be launched ?
    It still remains to me unclear, and did'nt yet investigate, on it.

  - Was developped (and so tested) on Cuda 9.2/PyTorch 0.4.1 (single and multi GPU)
    and also quickly cheked on a single GPU Cuda 10/PyTorch 1.0 Nightly build.
    CPU only, have barely not been tested at this point.

  - Dockers cpu and gpu have not been tested at all (cf pillow-simd stuff)

  - Didn't yet find an easy/efficient way to deal with PNG Palette with OpenCV2.
    Could leads to remove Pillow, as OpenCV2 is faster.

  - Code concision is kept with less than ~400 additionnal lines, in the codebase, for this whole PR ^^

  - Thanks in advance Daniel for the coming code review :)

### CODE REVIEW

**Comment by daniel-j-h** [comment]
Incredible! I need a few days to come back here and to go through this; putting it on my backlog.

**Comment by ocourtin** [comment]
@daniel-j-h 
OK, ping me when you will be available to begin the review.
In the meantime i keep working and pushing related stuff in this branch.

File: robosat/tools/download.py [hunk 1]
```diff
@@ -8,29 +8,42 @@
...
+        "download", help="downloads images from a remote server", formatter_class=argparse.ArgumentDefaultsHelpFormatter
     )
 
-    parser.add_argument("url", type=str, help="endpoint with {z}/{x}/{y} variables to fetch image tiles from")
+    parser.add_argument(
+        "url", type=str, help="endpoint with {z}/{x}/{y} or {xmin},{ymin},{xmax},{ymax} variables to fetch image tiles"
+    )
     parser.add_argument("--ext", type=str, default="webp", help="file format to save images in")
     parser.add_argument("--rate", type=int, default=10, help="rate limit in max. requests per second")
+    parser.add_argument("--type", type=str, default="XYZ", help="service type to use (e.g: XYZ, WMS or TMS)")
```
Comment (daniel-j-h) [inline robosat/tools/download.py]: Maybe add `choices`; then register handlers for the choices and have them download the corresponding files?

File: robosat/tools/export.py [hunk 2]
```diff
@@ -1,40 +1,60 @@
...
     net = torch.nn.DataParallel(net)
     net.load_state_dict(chkpt["state_dict"])
 
-    # Todo: make input channels configurable, not hard-coded to three channels for RGB
-    batch = torch.autograd.Variable(torch.randn(1, 3, args.image_size, args.image_size))
+    if export_channels < num_channels:
+        weights = torch.zeros((64, export_channels, 7, 7))
+        weights.data = net.module.resnet.conv1.weight.data[:, :export_channels, :, :]
+        net.module.resnet.conv1 = nn.Conv2d(num_channels, 64, kernel_size=7, stride=2, padding=3, bias=False)
+        net.module.resnet.conv1.weight = nn.Parameter(weights)
```
Comment (daniel-j-h) [inline robosat/tools/export.py]: Hm, this is a bit of a hack and breaks the model's class encapsulation. But on the other hand I don't see an elegant way...

File: robosat/tools/predict.py [hunk 3]
```diff
@@ -34,47 +36,47 @@ def add_parser(subparser):
...
+    num_classes = len(dataset["common"]["classes"])
 
-    cuda = model["common"]["cuda"]
-
-    device = torch.device("cuda" if cuda else "cpu")
+    if torch.cuda.is_available():
+        device = torch.device("cuda")
+        torch.backends.cudnn.benchmark = True
+    else:
+        device = torch.device("cpu")
```
Comment (daniel-j-h) [inline robosat/tools/predict.py]: :+1: I'm fine with switching the logic here to always using cuda if available. I think users still have the cuda environment variables they can set if they really don't want to use gpus but they do have gpus available.

File: robosat/tools/templates/leaflet.html [hunk 4]
```diff
@@ -0,0 +1,22 @@
+<!DOCTYPE html>
```
Comment (daniel-j-h) [inline robosat/tools/templates/leaflet.html]: We already have a map client with

https://github.com/mapbox/robosat/blob/master/robosat/tools/templates/map.html

using Mapbox GL JS where we can attach raster layer endpoints to.

What is the reason for pulling in a Leaflet based solution?

File: robosat/tools/tile.py [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+import numpy as np
+from PIL import Image
+
+import mercantile
+
+from rasterio import open as rasterio_open
+from rasterio.vrt import WarpedVRT
+from rasterio.enums import Resampling
+from rasterio.warp import transform_bounds, calculate_default_transform
+from rasterio.transform import from_bounds
```
Comment (daniel-j-h) [inline robosat/tools/tile.py]: Is there a reason we can't simply use the rio-tiler plugin?

File: robosat/tools/tile.py [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+
+        C, W, H = data.shape
+
+        os.makedirs(os.path.join(args.out, str(args.zoom), str(tile.x)), exist_ok=True)
+        path = os.path.join(args.out, str(args.zoom), str(tile.x), str(tile.y))
+
+        if args.type == "label":
+            assert C == 1, "Error: Label raster input should be 1 band"
+
+            ext = "png"
```
Comment (daniel-j-h) [inline robosat/tools/tile.py]: Should this be configurable?

File: robosat/tools/train.py [hunk 7]
```diff
@@ -247,34 +261,34 @@ def validate(loader, num_classes, device, net, criterion):
...
         [
-            JointTransform(ConvertImageMode("RGB"), ConvertImageMode("P")),
-            JointTransform(Resize(target_size, Image.BILINEAR), Resize(target_size, Image.NEAREST)),
-            JointTransform(CenterCrop(target_size), CenterCrop(target_size)),
-            JointRandomHorizontalFlip(0.5),
-            JointRandomRotation(0.5, 90),
-            JointRandomRotation(0.5, 90),
-            JointRandomRotation(0.5, 90),
+            JointResize(model["common"]["image_size"]),
+            JointRandomFlipOrRotate(model["opt"]["data_augmentation"]),
```
Comment (daniel-j-h) [inline robosat/tools/train.py]: :+1:

File: robosat/unet.py [hunk 8]
```diff
@@ -79,22 +80,30 @@ class UNet(nn.Module):
...
-        # Todo: make input channels configurable, not hard-coded to three channels for RGB
-
         self.resnet = resnet50(pretrained=pretrained)
 
-        # Access resnet directly in forward pass; do not store refs here due to
-        # https://github.com/pytorch/pytorch/issues/8392
+        assert num_channels
+
+        if num_channels != 3:
+            weights = nn.init.xavier_uniform_(torch.zeros((64, num_channels, 7, 7)))
```
Comment (daniel-j-h) [inline robosat/unet.py]: Did you test this? What I read was folks just zero-ing out the additional channels.

File: robosat/unet.py [hunk 9]
```diff
@@ -79,22 +80,30 @@ class UNet(nn.Module):
...
+            if pretrained:
+                for c in range(min(num_channels, 3)):
+                    weights.data[:, c, :, :] = self.resnet.conv1.weight.data[:, c, :, :]
+            self.resnet.conv1 = nn.Conv2d(num_channels, 64, kernel_size=7, stride=2, padding=3, bias=False)
+            self.resnet.conv1.weight = nn.Parameter(weights)
+
+        # No encoder reference, give a look at https://github.com/pytorch/pytorch/issues/8392
```
Comment (daniel-j-h) [inline robosat/unet.py]: It would be great to get this into https://github.com/mapbox/robosat/pull/75 too. And maybe we should get the FPN into master already. For my use-cases I'm seein much better runtime wrt. training and prediction and the results are almost the same to a bit better.

File: robosat/utils.py [hunk 10]
```diff
@@ -20,3 +25,23 @@ def plot(out, history):
 
     plt.savefig(out, format="png")
     plt.close()
+
+
+def leaflet(out, base_url, tiles, grid, ext):
+
+    grid = json.dumps([mercantile.feature(tile) for tile in grid]) if grid else "''"
+
+    leaflet = open("./robosat/tools/templates/leaflet.html", "r").read()
```
Comment (daniel-j-h) [inline robosat/utils.py]: This works only when invoked from within the robosat directory, no?

File: config/dataset-parking.toml [hunk 11]
```diff
@@ -21,3 +21,11 @@
...
 [weights]
   values = [1.6248, 5.762827]
+
+
+# Channels configuration let your indicate wich dataset sub-directory and bands to take as input
+# You could so, add several channels blocks to compose your input Tensor. Orders are meaningful.
+[[channels]]
+type = "file"
+sub = "images"
+bands = [1,2,3]
```
[the remark this answers is not in the document]
Comment (ocourtin) [reply config/dataset-parking.toml]: By convention raster bands are named 1-N (at the very least in GDAL and rasterio)

File: config/model-unet.toml [hunk 12]
```diff
@@ -32,3 +29,6 @@
 
   # Loss function name (e.g 'Lovasz', 'mIoU' or 'CrossEntropy')
   loss = 'Lovasz'
+
+  # Data augmentation, Flip or Rotate probability
+  data_augmentation = 0.75
```
[the remark this answers is not in the document]
Comment (ocourtin) [reply config/model-unet.toml]: The ideal value (on the dataset i play with) is related to the number of training epochs.

With a quick train, best result is with small or even no data augmentation.
But with long train, w/o data augmentation it's lead to overfit...

So it's looks like something a power user could want to change.

File: deps/requirements-lock.txt [hunk 13]
```diff
@@ -20,15 +20,15 @@ more-itertools==4.2.0
 numpy==1.14.4
 opencv-contrib-python==3.4.1.15
 osmium==2.14.1
-Pillow==5.1.0
+Pillow-simd==5.3.0
```
[the remark this answers is not in the document]
Comment (ocourtin) [reply deps/requirements-lock.txt]: In the train tool, indeed the workers performs well, and it's easy to feed fully the GPUs.
But with  predict, compare, or tile for instance, the simd switch is really impressive.

File: robosat/tools/tile.py (hunk 6 above)
Comment (ocourtin) [reply robosat/tools/tile.py]: Not sure to understand what do you have in mind on this one,

File: robosat/unet.py (hunk 8 above)
Comment (ocourtin) [reply robosat/unet.py]: I didn't check the difference between xavier or np.zeros.
But you're right it's worth to.
I look at it.

File: robosat/utils.py (hunk 10 above)
Comment (ocourtin) [reply robosat/utils.py]: Interesting question !
As i didn't find anything related to setup/deploy so far, i still launch rs commands from the robosat directory.

Is there already an another way to do ?

**Comment by ocourtin** [comment]
So any new comments welcome :)

**Comment by DavidDohmen** [comment]
Hi @ocourtin and @daniel-j-h I'm glad to see this amazing merge request! I wanted to ask if there's any progress on this and if there are any blockers left for a merge? The discussion seems to be stalled.

**Comment by ocourtin** [comment]
Hi @DavidDohmen !
Current dev, including this PR, and lot more enhancements, continue right now on RoboSat.pink fork.

GitHub: https://github.com/datapink/robosat.pink        (use the master till the coming 0.4.0 release)
Gitter Chat: https://gitter.im/RoboSatPink


I let @daniel-j-h answer if there's a way to resume devs in this repo.

**Comment by DavidDohmen** [comment]
Thanks for your quick answer and pointing me in the right direction! I will look into this!
